### PR TITLE
Add LoRA to BLOOM

### PR DESCRIPTION
**Pull request by haileyschoelkopf**
This PR adds LoRA to adapter-transformers' BLOOM implementation.

I have verified that with no LoRA adapter added to the model (just using `pfeiffer+inv` config for adapters), the performance is *exactly* the same (with a seed set) as before I added this change.

so this can be safely merged for us to try out.

TODO: 
- test `madx_run_clm.py` using a LoRA adapter (+ determine appropriate config values.)

### CODE REVIEW

File: src/transformers/models/bloom/modeling_bloom.py [hunk 1]
```diff
@@ -397,6 +400,8 @@ def forward(
             present = (key_layer, value_layer)
         else:
             present = None
+            
+        key_layer, value_layer, attention_mask = self.prefix_tuning(key_layer, value_layer, attention_mask)
```
Comment (yongzx) [inline src/transformers/models/bloom/modeling_bloom.py]: You need to transpose the `key_layer` and `value_layer` here before passing into prefix_tuning.

Comment (yongzx) [reply src/transformers/models/bloom/modeling_bloom.py]: Quick script to check if forward pass works:
```
from transformers import AutoModel, AutoTokenizer
model = AutoModel.from_pretrained("bigscience/bloom-1b3", cache_dir="/users/zyong2/data/zyong2/huggingface/", use_auth_token=True)
tokenizer = AutoTokenizer.from_pretrained("bigscience/bloom-1b3", cache_dir="/users/zyong2/data/zyong2/huggingface/", use_auth_token=True)

# model = AutoModel.from_pretrained("gpt2", cache_dir="/users/zyong2/data/zyong2/huggingface/")
# tokenizer = AutoTokenizer.from_pretrained("gpt2", cache_dir="/users/zyong2/data/zyong2/huggingface/")

sentence = "It's also, clearly, great fun."
input_data = tokenizer(sentence, return_tensors='pt')

from transformers.adapters import PrefixTuningConfig

config = PrefixTuningConfig(flat=False, prefix_length=30)
model.add_adapter("prefix_tuning", config=config)
model.train_adapter("prefix_tuning")

# print(model)
# 
outputs = model(**input_data)
print(outputs)
```

**Comment by yongzx** [comment]
@haileyschoelkopf can you separate prefix tuning from this PR (by creating another PR)? 

I have tested LoRA and the implementation looks correct! Once I have obtained the training curves from `madx_run_clm.py` (currently running), I would just merge this PR for LoRA.

**Comment by haileyschoelkopf** [comment]
I've just done this! Will open a separate PR for prefix tuning
